### ISSUE TITLE
recursor: linking errors in YaHTTP::AsyncLoader when building with gcc-10

### ISSUE THREAD

**Issue by sergeev917** [PowerDNS/pdns#9139]
- Program: Recursor
 - Issue type: Bug report

### Short description
PowerDNS Recursor fails to link because of issues in YaHTTP sources. For further details see "Other information".

### Environment
 - Operating system: Gentoo Linux
 - Software version: PowerDNS Recursor 4.3.1
 - Software source: Distribution repository (building from source)

### Steps to reproduce
1. Build PowerDNS Recursor 4.3.1
2. Observe a failure during linking

### Expected behaviour
The linking stage shouldn't have failed.

### Actual behaviour
The link cmd line and resulting errors as follows:
```
libtool: link: x86_64-pc-linux-gnu-g++ -DSYSCONFDIR=\"/etc/powerdns\" -DPKGLIBDIR=\"/usr/lib64/pdns-recursor\" -DLOCALSTATEDIR=\"/var/run\" -DNODCACHEDIR=\"/var/lib/lib/pdns-recursor\" -fPIE -DPIE -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 --param ssp-buffer-size=4 -fstack-protector -Wall -Wextra -Wshadow -Wno-unused-parameter -g -O2 -O2 -pipe -march=native -m64 -pie -pthread -Wl,-z -Wl,relro -Wl,-z -Wl,now -Wl,-O1 -Wl,--as-needed -o pdns_recursor arguments.o base32.o base64.o capabilities.o dns.o dns_random.o dnslabeltext.o dnsname.o dnsparser.o dnsrecords.o dnssecinfra.o dnswriter.o ednsoptions.o ednssubnet.o filterpo.o fstrm_logger.o gettime.o gss_context.o iputils.o ixfr.o json.o logger.o lua-base4.o lua-recursor4.o lwres.o misc.o mtasker_context.o negcache.o nsecrecords.o opensslsigners.o pdns_recursor.o pollmplexer.o protobuf.o pubsuffix.o pubsuffixloader.o qtype.o rcpgenerator.o rec-carbon.o rec-lua-conf.o rec-protobuf.o rec-snmp.o rec_channel.o rec_channel_rec.o recpacketcache.o recursor_cache.o reczones.o remote_logger.o resolver.o responsestats.o rpzloader.o secpoll-recursor.o secpoll.o sillyrecords.o snmp-agent.o sortlist.o syncres.o threadname.o tsigverifier.o unix_utility.o uuid-utils.o validate.o validate-recursor.o version.o webserver.o ws-api.o ws-recursor.o xpf.o zoneparser-tng.o nod.o sodiumsigners.o epollmplexer.o  ./ext/yahttp/yahttp/.libs/libyahttp.a ./ext/json11/.libs/libjson11.a -lcrypto -lboost_context-mt -lboost_system-mt ./ext/probds/.libs/libprobds.a -lcap -lboost_filesystem-mt -lsodium //usr/lib64//liblua.so -ldl -lm -lpthread -pthread -Wl,-rpath -Wl,/usr/lib64/ -Wl,-rpath -Wl,usr/lib64/
/usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: webserver.o: in function `WebServer::serveConnection(std::shared_ptr<Socket>) const':
/var/tmp/portage/net-dns/pdns-recursor-4.3.1/work/pdns-recursor-4.3.1/webserver.cc:384: undefined reference to `YaHTTP::AsyncLoader<YaHTTP::Request>::feed(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/lib/gcc/x86_64-pc-linux-gnu/10.1.0/../../../../x86_64-pc-linux-gnu/bin/ld: ws-recursor.o: in function `AsyncWebServer::serveConnection(std::shared_ptr<Socket>) const':
/var/tmp/portage/net-dns/pdns-recursor-4.3.1/work/pdns-recursor-4.3.1/ws-recursor.cc:694: undefined reference to `YaHTTP::AsyncLoader<YaHTTP::Request>::feed(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
```

### Other information
The following line declares `feed` method in `AsyncLoader` class template. Note that there is no inline definition of the method.
https://github.com/PowerDNS/pdns/blob/rec-4.3.1/ext/yahttp/yahttp/reqresp.hpp#L319

The definition is provided in the source file:
https://github.com/PowerDNS/pdns/blob/rec-4.3.1/ext/yahttp/yahttp/reqresp.cpp#L38

Classes `AsyncResponseLoader` and `AsyncRequestLoader` inherit from the `AsyncLoader` template instantiations with `T=Response` and `T=Request`.
https://github.com/PowerDNS/pdns/blob/rec-4.3.1/ext/yahttp/yahttp/reqresp.hpp#L344

The problem is that there is no `feed` method body provided for `AsyncLoader<T>::feed`, meaning that the compiler cannot generate code for `AsyncResponseLoader::feed` and `AsyncRequestLoader::feed`. The method definition in `reqresp.cpp` is of no use, it's a method of a class template -- the compiler cannot emit any code for it, since in this particular TU the compiler doesn't know about instantiations for `Response` and `Request`

Off the top of my head I could think of some ways of fixing this:
1) put the method definition into the header file;
2) similar to (1) -- `#include` the source file in the header file;
3) make use of explicit template instantiations, ie put `extern template class AsyncLoader<Response>;` to the header file and `template class AsyncLoader<Response>;` to the source file (same for `Request`).

I've tried the third option which fixed the problem for me.

**Comment by rgacogne** [comment]
This should have been fixed by https://github.com/PowerDNS/pdns/pull/9122. Unfortunately we forgot to do the backport before 4.3.1 so it will only be in 4.3.2 :-/

**Comment by rgacogne** [comment]
Fixed in 4.3.2.
